### PR TITLE
Rewrite send-trace-to-jaeger in Rust

### PR DESCRIPTION
**Pull request by timneutkens**
Just  using this as an opportunity to learn Rust given that it's an internal script :+1:

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

File: scripts/send-trace-to-jaeger/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "send-trace-to-jaeger"
+version = "0.1.0"
+edition = "2021"
+
```
Comment (kdy1) [inline scripts/send-trace-to-jaeger/Cargo.toml:5]: I think it's good to have `publish = false`.

File: scripts/send-trace-to-jaeger/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                };
+
+                let mapped = v
+                    .iter()
+                    .map(|data| {
+                        if !logged_url {
+                            log_web_url(&jaeger_web_ui_url, &data["traceId"].to_string());
+                            logged_url = true;
+                        }
+                        let mut data = data.clone();
```
Comment (kdy1) [inline scripts/send-trace-to-jaeger/src/main.rs:77]: This `clone` is not necessary and can be removed.

File: scripts/send-trace-to-jaeger/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+fn log_web_url(jaeger_web_ui_url: &str, trace_id: &str) {
+    println!(
+        "Jaeger trace will be available on {}/trace/{}",
+        jaeger_web_ui_url.to_string(),
+        trace_id.to_string()
+    )
+}
+
+/// Send trace JSON to Jaeger using ZipKin API.
+fn send_json_to_zipkin(zipkin_api: &str, value: &str) {
```
Comment (kdy1) [inline scripts/send-trace-to-jaeger/src/main.rs:27]: I think it's better to accept `value: String`, as this function unconditionally converts `value` to `String`.

**Review by kdy1 — COMMENTED** [review]
A new rustacean :)

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| buildDuration | 17.8s | 17.9s | ⚠️ +120ms |
| buildDurationCached | 3.6s | 3.5s | -24ms |
| nodeModulesSize | 335 MB | 335 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.891 | 2.944 | ⚠️ +0.05 |
| / avg req/sec | 864.81 | 849.27 | ⚠️ -15.54 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.364 | 1.427 | ⚠️ +0.06 |
| /error-in-render avg req/sec | 1833.12 | 1751.41 | ⚠️ -81.71 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.37 kB | 1.37 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.44 kB | 4.44 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.3 kB | 13.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 534 B | 534 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| buildDuration | 18.8s | 19s | ⚠️ +204ms |
| buildDurationCached | 3.7s | 3.5s | -136ms |
| nodeModulesSize | 335 MB | 335 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.873 | 2.847 | -0.03 |
| / avg req/sec | 870.15 | 877.97 | +7.82 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.377 | 1.406 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1815.2 | 1778.67 | ⚠️ -36.53 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.5 kB | 28.5 kB | ✓ |
| webpack-HASH.js gzip | 1.43 kB | 1.43 kB | ✓ |
| Overall change | 72.5 kB | 72.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.46 kB | 4.46 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.2 kB | 13.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 535 B | 535 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
Commit: 278db2d66dd9552ea721ca6808c02cd2b63c1e3d

File: scripts/send-trace-to-jaeger/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    );
+    local_endpoint.insert("ipv4".to_string(), Value::String(ipv4.to_string()));
+    local_endpoint.insert("port".to_string(), Value::Number(Number::from(port)));
+
+    let first_arg = args().nth(1).expect("Please provide a file name");
+
+    if let Ok(lines) = read_lines(first_arg) {
+        for line in lines {
+            if let Ok(json_to_parse) = line {
+                let v: Value = match serde_json::from_str::<Vec<Value>>(&json_to_parse) {
```
Comment (kdy1) [inline scripts/send-trace-to-jaeger/src/main.rs:62]: This type can be inferred by `rustc`, but it's simply matter of preference.

File: scripts/send-trace-to-jaeger/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                        .into_iter()
+                        .map(|mut data| {
+                            if !logged_url {
+                                log_web_url(&jaeger_web_ui_url, &data["traceId"].to_string());
+                                logged_url = true;
+                            }
+                            data["localEndpoint"] = Value::Object(local_endpoint.clone());
+                            data
+                        })
+                        .collect::<Value>(),
```
Comment (kdy1) [inline scripts/send-trace-to-jaeger/src/main.rs:73]: Alternatively, you can change this to `.collect()`

**Review by kdy1 — DISMISSED** [review]
It's simply a coding style issue, so I think it's now fine.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| buildDuration | 24.9s | 24.8s | -66ms |
| buildDurationCached | 4.9s | 4.8s | -112ms |
| nodeModulesSize | 335 MB | 335 MB | ⚠️ +1 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.217 | 4.194 | -0.02 |
| / avg req/sec | 592.85 | 596.06 | +3.21 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.301 | 2.183 | -0.12 |
| /error-in-render avg req/sec | 1086.27 | 1144.95 | +58.68 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.37 kB | 1.37 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.44 kB | 4.44 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.3 kB | 13.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 534 B | 534 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| buildDuration | 27s | 26.8s | -132ms |
| buildDurationCached | 4.8s | 5s | ⚠️ +164ms |
| nodeModulesSize | 335 MB | 335 MB | ⚠️ +1 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.195 | 4.204 | ⚠️ +0.01 |
| / avg req/sec | 595.95 | 594.72 | ⚠️ -1.23 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.28 | 2.226 | -0.05 |
| /error-in-render avg req/sec | 1096.42 | 1122.96 | +26.54 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.5 kB | 28.5 kB | ✓ |
| webpack-HASH.js gzip | 1.43 kB | 1.43 kB | ✓ |
| Overall change | 72.5 kB | 72.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.46 kB | 4.46 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.2 kB | 13.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | timneutkens/next.js add/rust-based-trace-import | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 535 B | 535 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
Commit: 7f0e2d9d8c3532d75c7c93d713a008351125c190

**Pull request merged**